components/feature/search/src/main/java/mozilla/components/feature/search/widget/BaseVoiceSearchActivity.kt

```diff
@@ -0,0 +1,132 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.search.widget
+
+import android.app.Activity
+import android.content.Intent
+import android.os.Bundle
+import android.speech.RecognizerIntent
+import androidx.activity.result.ActivityResult
+import androidx.activity.result.ActivityResultLauncher
+import androidx.activity.result.contract.ActivityResultContracts
+import androidx.annotation.VisibleForTesting
+import androidx.appcompat.app.AppCompatActivity
+import java.util.Locale
+
+/**
+ * Launches voice recognition then uses it to start a new web search.
+ */
+abstract class BaseVoiceSearchActivity : AppCompatActivity() {
+
+    /**
+     * Holds the intent that initially started this activity
+     * so that it can persist through the speech activity.
+     */
+    private var previousIntent: Intent? = null
+
+    @VisibleForTesting
+    private var activityResultLauncher: ActivityResultLauncher<Intent>? = null
+
+    override fun onSaveInstanceState(outState: Bundle) {
+        super.onSaveInstanceState(outState)
+        outState.putParcelable(PREVIOUS_INTENT, previousIntent)
+    }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        activityResultLauncher = getActivityResultLauncher()
+        if (Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH).resolveActivity(packageManager) == null) {
+            finish()
+            return
+        }
+
+        // Retrieve the previous intent from the saved state
+        previousIntent = savedInstanceState?.get(PREVIOUS_INTENT) as Intent?
+        if (previousIntent.isForSpeechProcessing()) {
+            // Don't reopen the speech recognizer
+            return
+        }
+
+        // The intent property is nullable, but the rest of the code below assumes it is not.
+        val intent = intent?.let { Intent(intent) } ?: Intent()
+        if (intent.isForSpeechProcessing()) {
+            previousIntent = intent
+            displaySpeechRecognizer()
+        } else {
+            finish()
+        }
+    }
+
+    /**
+     * Language locale for Voice Search
+     */
+    abstract fun getCurrentLocale(): Locale
+
+    /**
+     *records telemetry when speech recognizer popup is shown
+     */
+    abstract fun recordVoiceButtonTelemetry()
```

Comment:
nit: This method forces children to record telemetry.
Allowing children to know of certain events and then deciding for themselves if they want to log telemetry or not seems more suitable.
eg: This could be:
```kotlin
fun onSpeechRecognitionStarted() = {}
```

Reply:
Done